### DISCUSSION TITLE
[i18n] Allow having no default locale for international sites where all languages are equal

### DISCUSSION THREAD

**Discussion by advancingu** [vercel/next.js#18834] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

The existing internationalized routing mechanism in Next.js 10.0 does not allow building sites with dynamic content where no locale is considered a primary (default) locale.

For example, in https://nextjs.org/docs/advanced-features/i18n-routing#sub-path-routing , `en-US` is considered the primary locale that serves pages under `/` whereas all additional locales have pages relegated to locale-specific sub-paths such as `/nl-NL`.

For certain business cases however, no single locale must be considered primary and therefore all pages must reside in their locale-specific sub-paths. The unique remaining root path `/` will then perform automatic redirection of visitors to locale-specific pages based on business criteria (e.g. `Accept-Language` header, geo IP, etc.). No other routes must exist that are not locale-specific.

It is not possible to implement such a scenario with internationalized routing in Next.js 10.0.

## Describe the solution you'd like

A viable solution would be to allow setting `i18n.defaultLocale` to `null`. Consequently no routes under `/` must be generated other than the defined locales.

## Describe alternatives you've considered

An alternative is to manually build a folder structure that implements the locale sub-path via dynamic routes, e.g. `/[locale]/[slug].js`. This approach works, however features like automatic locale injection into the `Link` component have to be built manually.

## Additional context

This issue is highly relevant for sites generated predominantly from dynamic content (e.g. headless CMS) where page instances are generated via `getStaticPaths()`.

**Comment by omgMath** [comment]
Relates to (or duplicates?) https://github.com/vercel/next.js/discussions/18419

**Comment by advancingu** [comment]
@omgMath The issue itself describes a similar outcome, however its comments center around having redirects from a non-prefixed default locale path to a prefixed locale path corresponding to the default locale. The approach suggested in this issue is different in that it proposes that no redirects are generated.
